### ISSUE TITLE
False negative sometimes of `pub use module::Trait` to `pub(crate) use module::Trait`

### ISSUE THREAD

**Issue by leighmcculloch** [obi1kenobi/cargo-semver-checks#629]
### Steps to reproduce the bug with the above code

```
$ git clone https://github.com/stellar/rs-soroban-sdk

$ cd rs-soroban-sdk

$ git checkout 9f955efc767e145a200d804e11ba95f0bbff8634

$ sed -i.bak 's/pub use env::TryIntoVal;/pub(crate) use env::TryIntoVal;/' soroban-sdk/src/lib.rs

$ git diff
diff --git a/soroban-sdk/src/lib.rs b/soroban-sdk/src/lib.rs
index 458805b..fd1c0a0 100644
--- a/soroban-sdk/src/lib.rs
+++ b/soroban-sdk/src/lib.rs
@@ -740,7 +740,7 @@ pub use env::IntoVal;
 /// Used to do conversions between values in the Soroban environment.
 pub use env::TryFromVal;
 /// Used to do conversions between values in the Soroban environment.
-pub use env::TryIntoVal;
+pub(crate) use env::TryIntoVal;
 
 // Used by generated code only.
 #[doc(hidden)]

$ cargo semver-checks
...
```

### Actual Behaviour

```
$ cargo semver-checks
     Parsing soroban-ledger-snapshot v20.1.0 (current)
     Parsing soroban-ledger-snapshot v20.1.0 (baseline, cached)
    Checking soroban-ledger-snapshot v20.1.0 -> v20.1.0 (no change)
   Completed [   0.010s] 57 checks; 57 passed, 0 unnecessary
     Parsing soroban-sdk v20.1.0 (current)
     Parsing soroban-sdk v20.1.0 (baseline, cached)
    Checking soroban-sdk v20.1.0 -> v20.1.0 (no change)
   Completed [   0.122s] 57 checks; 57 passed, 0 unnecessary
     Parsing soroban-spec v20.1.0 (current)
     Parsing soroban-spec v20.1.0 (baseline, cached)
    Checking soroban-spec v20.1.0 -> v20.1.0 (no change)
   Completed [   0.008s] 57 checks; 57 passed, 0 unnecessary
     Parsing soroban-spec-rust v20.1.0 (current)
     Parsing soroban-spec-rust v20.1.0 (baseline, cached)
    Checking soroban-spec-rust v20.1.0 -> v20.1.0 (no change)
   Completed [   0.008s] 57 checks; 57 passed, 0 unnecessary
```

### Expected Behaviour

An error. Like what happens when I remove other traits. For example:
```
$ sed -i.bak 's/pub use env::IntoVal;/pub(crate) use env::IntoVal;/' soroban-sdk/src/lib.rs

$ git diff
diff --git a/soroban-sdk/src/lib.rs b/soroban-sdk/src/lib.rs
index 458805b..a0bf111 100644
--- a/soroban-sdk/src/lib.rs
+++ b/soroban-sdk/src/lib.rs
@@ -736,7 +736,7 @@ pub use env::Val;
 /// Used to do conversions between values in the Soroban environment.
 pub use env::FromVal;
 /// Used to do conversions between values in the Soroban environment.
-pub use env::IntoVal;
+pub(crate) use env::IntoVal;
 /// Used to do conversions between values in the Soroban environment.
 pub use env::TryFromVal;
 /// Used to do conversions between values in the Soroban environment.

$ cargo semver-checks
     Parsing soroban-ledger-snapshot v20.1.0 (current)
     Parsing soroban-ledger-snapshot v20.1.0 (baseline, cached)
    Checking soroban-ledger-snapshot v20.1.0 -> v20.1.0 (no change)
   Completed [   0.011s] 57 checks; 57 passed, 0 unnecessary
     Parsing soroban-sdk v20.1.0 (current)
     Parsing soroban-sdk v20.1.0 (baseline, cached)
    Checking soroban-sdk v20.1.0 -> v20.1.0 (no change)
   Completed [   0.117s] 57 checks; 55 passed, 2 failed, 0 unnecessary

--- failure trait_missing: pub trait removed or renamed ---

Description:
A publicly-visible trait cannot be imported by its prior path. A `pub use` may have been removed, or the trait itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.27.0/src/lints/trait_missing.ron

Failed in:
  trait soroban_sdk::IntoVal, previously in file /Users/leighmcculloch/.cargo/registry/src/index.crates.io-6f17d22bba15001f/soroban-sdk-20.1.0/src/env.rs:89

--- failure trait_removed_supertrait: supertrait removed or renamed ---

Description:
A supertrait was removed from a trait. Users of the trait can no longer assume it can also be used like its supertrait.
        ref: https://doc.rust-lang.org/reference/items/traits.html#supertraits
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.27.0/src/lints/trait_removed_supertrait.ron

Failed in:
  supertrait soroban_sdk::IntoVal of trait SorobanArbitrary in file /Users/leighmcculloch/Code/rs-soroban-sdk/soroban-sdk/src/testutils/arbitrary.rs:221
  supertrait soroban_sdk::IntoVal of trait Topics in file /Users/leighmcculloch/Code/rs-soroban-sdk/soroban-sdk/src/events.rs:59
       Final [   0.118s] semver requires new major version: 2 major and 0 minor checks failed
     Parsing soroban-spec v20.1.0 (current)
     Parsing soroban-spec v20.1.0 (baseline, cached)
    Checking soroban-spec v20.1.0 -> v20.1.0 (no change)
   Completed [   0.008s] 57 checks; 57 passed, 0 unnecessary
     Parsing soroban-spec-rust v20.1.0 (current)
     Parsing soroban-spec-rust v20.1.0 (baseline, cached)
    Checking soroban-spec-rust v20.1.0 -> v20.1.0 (no change)
   Completed [   0.008s] 57 checks; 57 passed, 0 unnecessary
```

#### Software version

cargo-semver-checks 0.27.0

#### Operating system

macOS 14.2.1 (Darwin 23.2.0)

#### Command-line

```bash
/Users/leighmcculloch/.cargo/bin/cargo-semver-checks semver-checks --bugreport 
```

#### cargo version

```
> cargo -V 
cargo 1.75.0 (1d8b05cdd 2023-11-20)
```

#### Compile time information

- Profile: release
- Target triple: aarch64-apple-darwin
- Family: unix
- OS: macos
- Architecture: aarch64
- Pointer width: 64
- Endian: little
- CPU features: aes,crc,dit,dotprod,dpb,dpb2,fcma,fhm,flagm,fp16,frintts,jsconv,lor,lse,neon,paca,pacg,pan,pmuv3,ras,rcpc,rcpc2,rdm,sb,sha2,sha3,ssbs,vh
- Host: aarch64-apple-darwin

### Build Configuration

None

**Comment by leighmcculloch** [comment]
The `TryIntoVal` type is defined in another crate that is then reexported and part of the API of that crate.

It's defined in the `soroban-env-common` crate, then reexported in the `soroban-env-guest` or `soroban-env-host` crates and imported into the `env` module of `soroban-sdk` then re-exported from there. The type is part of the public interface of the soroban-sdk crate though, you can see it in the docs.

I have noticed that when I break the API of other types like structs, such as the `Val` type which also lives in `soroban-env-common` and is reexported by `soroban-sdk`, that the semver checks is unable to identify that the API has broken.

**Comment by obi1kenobi** [comment]
Ah, that's the cause then.

`cargo-semver-checks` currently only does crate-local analysis. Any items from other crates are "as if" they don't exist, because they quite literally aren't present in the rustdoc JSON file that `cargo-semver-checks` sees. The "as if" includes re-exports of such items as well.

The limiting factors here are both in rustdoc itself and in `cargo-semver-checks`, and will require a few person-months to solve. Unfortunately, it's hard to dedicate so much time while also having to figure out how to pay my rent. It's a project I'd love to take on, and I'd appreciate your support! It could be in the form of a [GitHub Sponsors pledge](https://github.com/sponsors/obi1kenobi) which helps me maintain the tool _in general_, or in the form of sponsoring a fix for this issue _specifically_, for which I'd be happy to give you an estimate in time and money.

More details about the limitations can be found in this other issue: https://github.com/obi1kenobi/cargo-semver-checks/issues/573#issuecomment-1787288477

**Comment by obi1kenobi** [comment]
I sympathize with your concern. It would have been wonderful if cargo-semver-checks could have noticed and warned you there. I'm sorry it didn't!

Unfortunately the warning and a full fix are approximately the same difficulty to implement in practice.

The good news is that RustWeek and the associated Rust Project all hands meeting brought us much closer to unblocking this. But many person-months of work still remain, and cargo-semver-checks is still extremely short on funding. So the best way to move this forward is still to help us find more funding!
